subsys/bluetooth/gatt_dm.c

```diff
@@ -277,12 +295,29 @@ static void discovery_complete_error(struct bt_gatt_dm *dm, int err)
 	}
 }
 
+static struct gatt_discover_work_params {
+	struct k_work work;
+	struct bt_gatt_dm *dm;
```

Comment:
I'd recommend to move `struct k_work work` under `struct bt_gatt_dm`. I see that currently only single `bt_gatt_dm` instance is supported, but if it is solved later  this code may be broken.